data_model/src/evaluate.rs

```diff
@@ -1,299 +1,277 @@
 //! Implementations for Expression evaluation for different expressions.
-#![allow(
-    clippy::module_name_repetitions,
-    clippy::std_instead_of_core,
-    clippy::std_instead_of_alloc,
-    clippy::arithmetic_side_effects
-)]
-use eyre::Result;
-use iroha_data_model::{
+
+#[cfg(not(feature = "std"))]
+use alloc::{
+    collections::BTreeMap,
+    string::{String, ToString},
+};
+#[cfg(feature = "std")]
+use std::collections::BTreeMap;
+
+use derive_more::Display;
+use iroha_macro::FromVariant;
+
+use crate::{
     expression::{prelude::*, Expression},
-    isi::error::{InstructionExecutionFailure as Error, MathError},
+    isi::error::MathError,
+    model,
     prelude::*,
-    query::error::FindError,
+    query::error::QueryExecutionFailure,
 };
 
-use super::Evaluate;
-use crate::{prelude::ValidQuery, wsv::WorldStateView};
+/// Context of expression evaluation, holding (name, value) pairs for resolving identifiers.
+/// Context comes into play because of [`Where`] and [`Query`] expressions.
+///
+/// # Example
+///
+/// Say you have an expression such as: `SELECT name FROM table WHERE name = "alice"`. This
+/// compound expression is made up of two basic expressions, namely `SELECT FROM` and `WHERE`.
+/// To evaluate any expresion you have to substitute concrete values for variable names.
+/// In this case, `WHERE` should be evaluated first which would place `name = "alice"`
+/// inside the context. This context will then be used to evaluate `SELECT FROM`.
+/// Starting expression would then be evaluated to `SELECT "alice" FROM table`
+pub trait Context: Clone {
+    /// Execute query against the current state of `Iroha`
+    ///
+    /// # Errors
+    ///
+    /// If query execution fails
+    fn query(&self, query: &QueryBox) -> Result<Value, QueryExecutionFailure>;
+
+    /// Return a reference to the [`Value`] corresponding to the [`Name`].
+    fn get(&self, name: &Name) -> Option<&Value>;
+
+    /// Update this context with given values.
+    fn update(&mut self, other: impl IntoIterator<Item = (Name, Value)>);
+}
+
+/// Calculate the result of the expression without mutating the state.
+pub trait Evaluate {
+    /// The resulting type of the expression.
+    type Value;
```

Comment:
Isn't it always `iroha_data_model::Value`?

Reply:
no, `EvaluatesTo<V>` evaluates to a concrete type `V`